drivers/clock_control/nrf_power_clock.c

```diff
@@ -280,66 +322,64 @@ static int hfclk_init(struct device *dev)
 	}
 
 	clock_irqs_enable();
-	sys_slist_init(&((struct nrf_clock_control *)dev->driver_data)->list);
 
-	return 0;
-}
+	for (enum clock_control_nrf_type i = 0;
+		i < CLOCK_CONTROL_NRF_TYPE_COUNT; i++) {
+		sys_slist_init(&(get_sub_data(dev, i)->list));
+	}
 
-static int lfclk_init(struct device *dev)
-{
-	sys_slist_init(&((struct nrf_clock_control *)dev->driver_data)->list);
 	return 0;
 }
-
 static const struct clock_control_driver_api clock_control_api = {
 	.on = clock_start,
 	.off = clock_stop,
 	.async_on = clock_async_start,
 	.get_status = get_status,
 };
 
-static struct nrf_clock_control hfclk;
-static const struct nrf_clock_control_config hfclk_config = {
-	.start_tsk = NRF_CLOCK_TASK_HFCLKSTART,
-	.started_evt = NRF_CLOCK_EVENT_HFCLKSTARTED,
-	.stop_tsk = NRF_CLOCK_TASK_HFCLKSTOP
-};
-
-DEVICE_AND_API_INIT(clock_nrf5_m16src,
-		    DT_INST_0_NORDIC_NRF_CLOCK_LABEL  "_16M",
-		    hfclk_init, &hfclk, &hfclk_config, PRE_KERNEL_1,
-		    CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
-		    &clock_control_api);
-
-
-static struct nrf_clock_control lfclk;
-static const struct nrf_clock_control_config lfclk_config = {
-	.start_tsk = NRF_CLOCK_TASK_LFCLKSTART,
-	.started_evt = NRF_CLOCK_EVENT_LFCLKSTARTED,
-	.stop_tsk = NRF_CLOCK_TASK_LFCLKSTOP,
-	.start_handler =
-		IS_ENABLED(CONFIG_CLOCK_CONTROL_NRF_K32SRC_RC_CALIBRATION) ?
-			z_nrf_clock_calibration_start : NULL,
-	.stop_handler =
-		IS_ENABLED(CONFIG_CLOCK_CONTROL_NRF_K32SRC_RC_CALIBRATION) ?
-			z_nrf_clock_calibration_stop : NULL
+static struct nrf_clock_control_data data;
+
+static const struct nrf_clock_control_config config = {
+	.subsys = {
+		[CLOCK_CONTROL_NRF_TYPE_HFCLK] = {
+			.start_tsk = NRF_CLOCK_TASK_HFCLKSTART,
+			.started_evt = NRF_CLOCK_EVENT_HFCLKSTARTED,
+			.stop_tsk = NRF_CLOCK_TASK_HFCLKSTOP,
+			.name = IS_ENABLED(CONFIG_LOG) ? "hfclk" : NULL
```

Comment:
This field is unneeded when logging is disabled, so perhaps it would be better to wrap it in `#if IS_ENABLED(CONFIG_LOG)`?

Reply:
done